### DISCUSSION TITLE
What's the meaning of driver esp32sx?

### DISCUSSION THREAD

**Discussion by Huckies** [hathach/tinyusb#1565] [Q&A] [answered]
Here it says esp32-s series chips can use either the dwc2 driver or the esp32sx driver.
[https://github.com/hathach/tinyusb/blob/master/docs/reference/supported.rst](url)

But in fact that esp32sx driver has been desolated for few months while the dwc2 driver is updated frequently, so may I ask the significance? @hathach

**Comment by hathach** [comment, marked as answer]
similar question is answered in https://github.com/hathach/tinyusb/discussions/1546#discussioncomment-3169408

**Reply by Huckies** [reply]
The funny thing is, the esp32sx driver can lead to choked endpoint while dwc2 does not. If multiple endpoints can't be accessed by different tasks simultaneously, I'd rather use report ids.

Even more, a traffic jam can happen when feeding one endpoint per polling interval. Although you can defer this by raising the processor's clock frequency, it happens eventually.

**Reply by hathach** [reply]
then you could use dwc2 if that works for you. They are quite a bit difference, in general most recent fixes are applied to dwc2 only.
